packages/core/src/execution-engine/node-execution-context/utils/file-system-helper-functions.ts

```diff
@@ -74,34 +76,59 @@ export const getFileSystemHelperFunctions = (node: INode): FileSystemHelperFunct
 		}
 
 		try {
-			await fsAccess(filePath);
+			await fsAccess(resolvedFilePath);
 		} catch (error) {
 			// eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
 			throw error.code === 'ENOENT'
 				? // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
 					new NodeOperationError(node, error, {
-						message: `The file "${String(filePath)}" could not be accessed.`,
+						message: `The file "${String(resolvedFilePath)}" could not be accessed.`,
 						level: 'warning',
 					})
 				: error;
 		}
 
-		return createReadStream(filePath);
+		// Use O_NOFOLLOW to prevent createReadStream from following symlinks. We require that the path
+		// already be resolved beforehand.
+		const stream = createReadStream(resolvedFilePath, {
+			flags: (constants.O_RDONLY | constants.O_NOFOLLOW) as unknown as string,
+		});
+
+		return await new Promise<ReturnType<typeof createReadStream>>((resolve, reject) => {
+			stream.once('error', (error) => {
+				if ((error as NodeJS.ErrnoException).code === 'ELOOP') {
+					reject(
+						new NodeOperationError(
+							node,
+							`The file "${String(resolvedFilePath)}" could not be opened.`,
+							{ level: 'warning' },
+						),
+					);
+				} else {
+					reject(error);
+				}
+			});
+			stream.once('open', () => resolve(stream));
+		});
 	},
 
 	getStoragePath() {
 		return safeJoinPath(Container.get(InstanceSettings).n8nFolder, `storage/${node.type}`);
 	},
 
-	async writeContentToFile(filePath, content, flag) {
-		if (await isFilePathBlocked(filePath as string)) {
-			throw new NodeOperationError(node, `The file "${String(filePath)}" is not writable.`, {
-				level: 'warning',
-			});
+	async writeContentToFile(resolvedFilePath, content, flag) {
+		if (isFilePathBlocked(resolvedFilePath)) {
+			throw new NodeOperationError(
+				node,
+				`The file "${String(resolvedFilePath)}" is not writable.`,
+				{
+					level: 'warning',
+				},
+			);
 		}
-		return await fsWriteFile(filePath, content, { encoding: 'binary', flag });
+		return await fsWriteFile(resolvedFilePath, content, { encoding: 'binary', flag });
```

Comment:
Doesn't this still have the issue? Can we make this so that `fsWriteFile` doesn't follow symlinks?